pkg/controller/rds/dbparametergroup/setup.go

```diff
@@ -0,0 +1,150 @@
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.DBParameterGroupGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.RDSAPI
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.DescribeDBParameterGroupsInput) error {
+	obj.DBParameterGroupName = awsclients.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.DescribeDBParameterGroupsOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	cr.SetConditions(xpv1.Available())
+	return obs, err
+}
+
+// func postCreate(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.CreateDBParameterGroupOutput, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+// 	if err != nil {
+// 		return managed.ExternalCreation{}, err
+// 	}
+// 	meta.SetExternalName(cr, awsclients.StringValue(obj.DBParameterGroup.DBParameterGroupName))
+// 	return managed.ExternalCreation{ExternalNameAssigned: true}, nil
+// }
+
+func preCreate(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.CreateDBParameterGroupInput) error {
+	obj.DBParameterGroupName = awsclients.String(meta.GetExternalName(cr))
+
+	return nil
+}
+
+func preUpdate(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.ModifyDBParameterGroupInput) error {
+	obj.DBParameterGroupName = awsclients.String(meta.GetExternalName(cr))
+	obj.Parameters = make([]*rds.Parameter, len(cr.Spec.ForProvider.Parameters))
+
+	for i, v := range cr.Spec.ForProvider.Parameters {
+		// check if mandatory parameters are set (ApplyMethod, ParameterName, ParameterValue)
+		if (v.ApplyMethod == nil) || (v.ParameterName == nil) || (v.ParameterValue == nil) {
+			return errors.New("ApplyMethod, ParameterName and ParameterValue are mandatory fields and can not be nil")
+		}
+		obj.Parameters[i] = &rds.Parameter{
+			ApplyMethod:    awsclients.String(*v.ApplyMethod),
+			ParameterName:  awsclients.String(*v.ParameterName),
+			ParameterValue: awsclients.String(*v.ParameterValue),
+		}
+	}
+	return nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.DBParameterGroup, obj *svcsdk.DeleteDBParameterGroupInput) (bool, error) {
+	obj.DBParameterGroupName = awsclients.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+func (e *custom) isUpToDate(cr *svcapitypes.DBParameterGroup, obj *svcsdk.DescribeDBParameterGroupsOutput) (bool, error) {
+	// TODO(Dkaykay): We need isUpToDate to have context.
+	ctx := context.TODO()
+	results, err := e.getCurrentDBParameters(ctx, cr)
+
+	// compare CR with currently set Parameters
```

Comment:
Since it's possible for the parameters to be many, what do you think about using a `map[string]svcapitypes.Parameter`? First you would check the length of both arrays and then one loop would fill the map and the second loop would do the checks against fetched parameters by using `ParameterName` as the key.

Reply:
I think i dont get this one. The getCurrentDBParameters function returns an array of *svcsdk.Parameter, so all parameters currently set on AWS (about 300). In isUpToDate all parameters are compared with the ones set in the dbparametergroup claim

Reply:
It'd look like the following:
```golang
func (e *custom) isUpToDate(cr *svcapitypes.DBParameterGroup, obj *svcsdk.DescribeDBParameterGroupsOutput) (bool, error) {
	// TODO(Dkaykay): We need isUpToDate to have context.
	ctx := context.TODO()
	results, err := e.getCurrentDBParameters(ctx, cr)
	if err != nil {
		return false, err
	}
	observed := make(map[string]svcsdk.Parameter, len(results))
	for _, p := range results {
		observed[awsclients.StringValue(p.ParameterName)] = *p
	}

	// compare CR with currently set Parameters
	for _, v := range cr.Spec.ForProvider.Parameters {
		existing, ok := observed[awsclients.StringValue(v.ParameterName)]
		if !ok {
			return false, nil
		}
		switch {
		case awsclients.StringValue(existing.ParameterValue) != awsclients.StringValue(v.ParameterName):
			return false, nil
		case awsclients.StringValue(existing.ApplyMethod) != awsclients.StringValue(v.ApplyMethod):
			return false, nil
		}
	}
	return true, nil
}
```

Reply:
Gotcha! will do